monty-31/src/x86_64_avx512/poseidon2.rs

```diff
@@ -1,34 +1,450 @@
+/// Each entry of rc should be represented by an element in in [-P, 0].
+/// Each entry of the output will be represented by an element in canonical form.
+/// If the inputs do not conform to this representation, the result is undefined.
+#[inline(always)]
+fn add_rc_and_sbox<PMP: PackedMontyParameters, const D: u64>(
+    val: &mut PackedMontyField31AVX512<PMP>,
+    rc: __m512i,
+) {
+    unsafe {
+        // As our exponential functions simply assume that
+        // the input lies in [-P, P] we do not need to perform a reduction provided
+        // rc is represented by an element in [-P, 0]
+        let vec_val = val.to_vector();
+        let val_plus_rc = x86_64::_mm512_add_epi32(vec_val, rc);
+        let output = apply_func_to_even_odd::<PMP>(val_plus_rc, exp_small::<PMP, D>);
+
+        *val = PackedMontyField31AVX512::<PMP>::from_vector(output);
+    }
+}
+
+/// A trait containing the specific information needed to
+/// implement the Poseidon2 Permutation for Monty31 Fields.
+pub trait InternalLayerParametersAVX512<PMP: PackedMontyParameters, const WIDTH: usize>:
+    Clone + Sync
+{
+    type ArrayLike: AsMut<[__m512i]>;
+
+    // diagonal_mul and add_sum morally should be one function but are split because diagonal_mul can happen simultaneously to
+    // the sbox being applied to the first element of the state which is advantageous as this s-box has very high latency.
+    // However these functions should only ever be used together and we only make safety guarantees about the output
+    // of the combined function add_sum(diagonal_mul(state), sum) which will output field elements in canonical form provided inputs are in canonical form.
+
+    // Diagonal_mul will not output field elements in canonical form and indeed may even output incorrect values in places where
+    // it is efficient to pipe computation to add_sum. E.g. it might output 3*x instead of -3*x and then add_sum does sum - x.
+    // Similarly add_sum assumes its input has been piped directly from diagonal_mul so might assume that some inputs
+    // are the negative of the correct value or in some form other than canonical.
+
+    // For these reason we mark both functions as unsafe.
+
+    // All 4 implementation of this trait (Field = BabyBear/KoalaBear, WIDTH = 16/24) have a similarly structured
+    // diagonal matrix. The first 9 elements of this matrix are: [-2, 1, 2, 1/2, 3, 4, -1/2, -3, -4] the next few are
+    // positive inverse powers of two and the remainder are negative inverse powers of two.
+
+    /// # Safety
+    ///
+    /// This function assumes its output is piped directly into add_sum.
+    #[inline(always)]
+    unsafe fn diagonal_mul(input: &mut Self::ArrayLike) {
+        Self::diagonal_mul_first_eight(input);
+        Self::diagonal_mul_remainder(input);
+    }
+
+    /// # Safety
+    ///
+    /// This function assumes its output is piped directly into add_sum.
```

Comment:
Ah, clearly the output isn't piped into `add_sum`, as can be seen in the implementation of `diagonal_mul` above. Please update this comment (and the various places it is copypasted).